File: client-clojure/src/sirqul_io_t_platform/specs/bid_response.clj
(ns sirqul-io-t-platform.specs.bid-response
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            )
  (:import (java.io File)))


(def bid-response-data
  {
   (ds/opt :bidId) int?
   (ds/opt :biddableType) string?
   (ds/opt :biddableId) int?
   (ds/opt :amountPerView) float?
   (ds/opt :amountPerAction) float?
   (ds/opt :currentDailyBudget) float?
   (ds/opt :currentBudget) float?
   (ds/opt :currentBudgetExpiration) int?
   (ds/opt :currentBudgetStart) int?
   (ds/opt :budgetAmount) float?
   (ds/opt :budgetSchedule) string?
   })

(def bid-response-spec
  (ds/spec
    {:name ::bid-response
     :spec bid-response-data}))

File: client-scalaz/src/main/scala/org/openapitools/client/api/BidResponse.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime


import BidResponse._

case class BidResponse (
  bidId: Option[Long],
biddableType: Option[String],
biddableId: Option[Long],
amountPerView: Option[Double],
amountPerAction: Option[Double],
currentDailyBudget: Option[Double],
currentBudget: Option[Double],
currentBudgetExpiration: Option[Long],
currentBudgetStart: Option[Long],
budgetAmount: Option[Double],
budgetSchedule: Option[BudgetSchedule])

object BidResponse {
  import DateTimeCodecs._
  sealed trait BudgetSchedule
  case object DAILY extends BudgetSchedule
  case object WEEKLY extends BudgetSchedule
  case object MONTHLY extends BudgetSchedule
  case object YEARLY extends BudgetSchedule

  object BudgetSchedule {
    def toBudgetSchedule(s: String): Option[BudgetSchedule] = s match {
      case "DAILY" => Some(DAILY)
      case "WEEKLY" => Some(WEEKLY)
      case "MONTHLY" => Some(MONTHLY)
      case "YEARLY" => Some(YEARLY)
      case _ => None
    }

    def fromBudgetSchedule(x: BudgetSchedule): String = x match {
      case DAILY => "DAILY"
      case WEEKLY => "WEEKLY"
      case MONTHLY => "MONTHLY"
      case YEARLY => "YEARLY"
    }
  }

  implicit val BudgetScheduleEnumEncoder: EncodeJson[BudgetSchedule] =
    EncodeJson[BudgetSchedule](is => StringEncodeJson(BudgetSchedule.fromBudgetSchedule(is)))

  implicit val BudgetScheduleEnumDecoder: DecodeJson[BudgetSchedule] =
    DecodeJson.optionDecoder[BudgetSchedule](n => n.string.flatMap(jStr => BudgetSchedule.toBudgetSchedule(jStr)), "BudgetSchedule failed to de-serialize")

  implicit val BidResponseCodecJson: CodecJson[BidResponse] = CodecJson.derive[BidResponse]
  implicit val BidResponseDecoder: EntityDecoder[BidResponse] = jsonOf[BidResponse]
  implicit val BidResponseEncoder: EntityEncoder[BidResponse] = jsonEncoderOf[BidResponse]
}
